RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducerKernel.dev.cc

```diff
@@ -0,0 +1,1652 @@
+          // Divide by position norm
+          clusterPos.x /= clusterPos.w;
+          clusterPos.y /= clusterPos.w;
+          clusterPos.z /= clusterPos.w;
+
+          if (debug)
+            printf("\tPF cluster (seed %d) energy = %f\tposition = (%f, %f, %f)\n",
+                   i,
+                   clusterEnergy,
+                   clusterPos.x,
+                   clusterPos.y,
+                   clusterPos.z);
+        } else {
+          if (debug)
+            printf("\tPF cluster (seed %d) position norm (%f) less than minimum (%f)\n",
+                   i,
+                   clusterPos.w,
+                   pfClusParams.minAllowedNormalization());
+          clusterPos.x = 0.;
+          clusterPos.y = 0.;
+          clusterPos.z = 0.;
+        }
+        float diff2 = dR2(prevClusterPos, clusterPos);
+        if (debug)
+          printf("\tPF cluster (seed %d) has diff2 = %f\n", i, diff2);
+        prevClusterPos = clusterPos;  // Save clusterPos
+
+        float diff = sqrtf(diff2);
+        iter++;
+        notDone = (diff > tol) && (iter < pfClusParams.maxIterations());
+        if (debug) {
+          if (diff > tol)
+            printf("\tTopoId %d has diff = %f greater than tolerance %f (continuing)\n", topoId, diff, tol);
+          else if (debug)
+            printf("\tTopoId %d has diff = %f LESS than tolerance %f (terminating!)\n", topoId, diff, tol);
+        }
+      }
+      alpaka::syncBlockThreads(acc);  // all threads call sync
+    } while (notDone);                // shared variable condition ensures synchronization is well defined
+    if (tid == 0) {                   // Cluster is finalized, assign cluster information to te SoA
+      int rhIdx =
+          pfClusteringVars[pfClusteringVars[topoId].topoSeedOffsets()].topoSeedList();  // i is the seed rechit index
+      int seedIdx = pfClusteringVars[rhIdx].rhIdxToSeedIdx();
+      clusterView[seedIdx].energy() = clusterEnergy;
+      clusterView[seedIdx].x() = clusterPos.x;
+      clusterView[seedIdx].y() = clusterPos.y;
+      clusterView[seedIdx].z() = clusterPos.z;
+    }
+  }
+
+  // Processing clusters up to 100 seeds and 512 non-seed rechits using shared memory accesses
+  template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+  ALPAKA_FN_ACC static void dev_hcalFastCluster_multiSeedParallel(
+      const TAcc& acc,
+      reco::PFClusterParamsDeviceCollection::ConstView pfClusParams,
+      int topoId,   // from selection
+      int nSeeds,   // from selection
+      int nRHTopo,  // from selection
+      reco::PFRecHitDeviceCollection::ConstView pfRecHits,
+      reco::PFClusteringVarsDeviceCollection::View pfClusteringVars,
+      reco::PFClusterDeviceCollection::View clusterView,
+      reco::PFRecHitFractionDeviceCollection::View fracView) {
+    int tid = alpaka::getIdx<alpaka::Block,
+                             alpaka::Threads>(  // Thread index corresponds to a single rechit of the topo cluster
```

Comment:
probably works better to move the comment above the declaration, instead of to the side